(regression) SIGSEGV in RuntimeMethodInfo:InternalInvoke with xunit

Environment: 
Ubuntu 16.04 Azure Pipelines image with
`Mono JIT compiler version 5.18.0.268 (tarball Thu Feb 28 15:29:06 UTC 2019)`

was working completely fine with 
`Mono JIT compiler version 5.18.0.240 (tarball Wed Jan 16 09:13:03 UTC 2019)`


Repro:
```
git clone --recursive git@github.com:AvaloniaUI/Avalonia.git
cd Avalonia
sudo apt-get update && sudo apt-get install castxml
dotnet tool install --global Nuke.GlobalTool --version 0.12.3
nuke --target CiAzureLinux --configuration=Release
```

One of xunit test assemblies segfaults like this https://dev.azure.com/AvaloniaUI/AvaloniaUI/_build/results?buildId=1851&view=logs&jobId=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&taskId=d9b15392-e4ce-5e4c-0c8c-b69645229181&lineStart=1132&lineEnd=1133&colStart=1&colEnd=1


I've also checked with mono built from master, still segfaults, but I'm also seeing messages like these:
```
ves_icall_TypeBuilder_create_runtime_class_raw USED 151 handles
ves_icall_TypeBuilder_create_runtime_class_raw USED 156 handles
ves_icall_TypeBuilder_create_runtime_class_raw USED 104 handles
ves_icall_TypeBuilder_create_runtime_class_raw USED 108 handles
ves_icall_TypeBuilder_create_runtime_class_raw USED 158 handles
```

Not sure if that's relevant.




Also reproduced on OSX with `MonoFramework-MDK-5.18.0.268.macos10.xamarin.universal.pkg`, works fine with 5.18.0.225.

Must be handle creation in a loop.

I tried building Avalonia.
I could not get `nuke` to run and `build.sh` failed.

```
jay@jaykrell-tp2:/mnt/c/s/Avalonia$ sh -c ./build.sh
GNU bash, version 4.4.19(1)-release (x86_64-pc-linux-gnu)
Microsoft (R) .NET Core SDK version 2.2.203
Shims.cs(8,7): error CS0246: The type or namespace name 'Numerge' could not be found (are you missing a using directive or an assembly reference?) [/mnt/c/s/Avalonia/nukebuild/_build.csproj]
Shims.cs(81,31): error CS0246: The type or namespace name 'INumergeLogger' could not be found (are you missing a using directive or an assembly reference?) [/mnt/c/s/Avalonia/nukebuild/_build.csproj]
Shims.cs(83,25): error CS0246: The type or namespace name 'NumergeLogLevel' could not be found (are you missing a using directive or an assembly reference?) [/mnt/c/s/Avalonia/nukebuild/_build.csproj]

The build failed. Please fix the build errors and run again.
jay@jaykrell-tp2:/mnt/c/s/Avalonia$  nuke
nuke: command not found
jay@jaykrell-tp2:/mnt/c/s/Avalonia$ dotnet nuke
No executable found matching command "dotnet-nuke"
jay@jaykrell-tp2:/mnt/c/s/Avalonia$
```

--- a/mcs/class/corlib/Test/System.Reflection/MethodInfoTest.cs
+++ b/mcs/class/corlib/Test/System.Reflection/MethodInfoTest.cs
@@ -658,6 +658,15 @@ namespace MonoTests.System.Reflection
 			Assert.AreEqual (10, pi2.GetGetMethod ().Invoke (10, null));
 		}
 
+		[Test]
+		public void NullableTestsStatic ()
+		{
+			Nullable<Double> val = new Nullable<Double>(new Double());
+			MethodInfo mi = typeof (Nullable<Double>).GetMethod ("op_Implicit");
+            object obj = val;
+            mi.Invoke(null, new[] { obj });
+		}
+
 		public static void foo_generic<T> ()
 		{
 		}

--- a/mcs/class/corlib/Test/System.Reflection/MethodInfoTest.cs
+++ b/mcs/class/corlib/Test/System.Reflection/MethodInfoTest.cs
@@ -663,8 +663,8 @@ namespace MonoTests.System.Reflection
 		{
 			Nullable<Double> val = new Nullable<Double>(new Double());
 			MethodInfo mi = typeof (Nullable<Double>).GetMethod ("op_Implicit");
-            object obj = val;
-            mi.Invoke(null, new[] { obj });
+			object obj = val;
+			mi.Invoke(null, new[] { obj });
 		}
 
 		public static void foo_generic<T> ()

--- a/mono/metadata/object.c
+++ b/mono/metadata/object.c
@@ -5494,16 +5494,19 @@ mono_runtime_try_invoke_array (MonoMethod *method, void *obj, MonoArray *params,
 		return (MonoObject *)obj;
 	} else {
 		if (mono_class_is_nullable (method->klass)) {
-			MonoObject *nullable;
+			if (method->flags & METHOD_ATTRIBUTE_STATIC) {
+				obj = NULL;
+			} else {
+				MonoObject *nullable;
+				/* Convert the unboxed vtype into a Nullable structure */
+				nullable = mono_object_new_checked (mono_domain_get (), method->klass, error);
+				return_val_if_nok (error, NULL);
 
-			/* Convert the unboxed vtype into a Nullable structure */
-			nullable = mono_object_new_checked (mono_domain_get (), method->klass, error);
-			return_val_if_nok (error, NULL);
-
-			MonoObject *boxed = mono_value_box_checked (mono_domain_get (), m_class_get_cast_class (method->klass), obj, error);
-			return_val_if_nok (error, NULL);
-			mono_nullable_init ((guint8 *)mono_object_unbox_internal (nullable), boxed, method->klass);
-			obj = mono_object_unbox_internal (nullable);
+				MonoObject *boxed = mono_value_box_checked (mono_domain_get (), m_class_get_cast_class (method->klass), obj, error);
+				return_val_if_nok (error, NULL);
+				mono_nullable_init ((guint8 *)mono_object_unbox_internal (nullable), boxed, method->klass);
+				obj = mono_object_unbox_internal (nullable);
+			}
 		}
 
 		/* obj must be already unboxed if needed */
@@ -6894,6 +6897,7 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	error_init (error);
 
 	g_assert (m_class_is_valuetype (klass));
+	g_assert (value != NULL);
 	if (G_UNLIKELY (m_class_is_byreflike (klass))) {
 		char *full_name = mono_type_get_full_name (klass);
 		mono_error_set_execution_engine (error, "Cannot box IsByRefLike type %s", full_name);
@@ -6903,9 +6907,6 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	if (mono_class_is_nullable (klass))
 		return mono_nullable_box_handle (value, klass, error);
 
-	if (!value) {
-		return NULL_HANDLE;
-	}
 	
 	vtable = mono_class_vtable_checked (domain, klass, error);
 	return_val_if_nok (error, NULL_HANDLE);

--- a/mono/metadata/object.c
+++ b/mono/metadata/object.c
@@ -6903,6 +6903,10 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	if (mono_class_is_nullable (klass))
 		return mono_nullable_box_handle (value, klass, error);
 
+	if (!value) {
+		return NULL_HANDLE;
+	}
+	
 	vtable = mono_class_vtable_checked (domain, klass, error);
 	return_val_if_nok (error, NULL_HANDLE);
 
@@ -6912,39 +6916,39 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	return_val_if_nok (error, NULL_HANDLE);
 
 	size -= MONO_ABI_SIZEOF (MonoObject);
-	if (value != NULL) { 
-		if (mono_gc_is_moving ()) {
-			g_assert (size == mono_class_value_size (klass, NULL));
-			MONO_ENTER_NO_SAFEPOINTS;
-			gpointer data = mono_handle_get_data_unsafe (res_handle);
-			mono_gc_wbarrier_value_copy_internal (data, value, 1, klass);
-			MONO_EXIT_NO_SAFEPOINTS;
-		} else {
-			MONO_ENTER_NO_SAFEPOINTS;
-			gpointer data = mono_handle_get_data_unsafe (res_handle);
-	#if NO_UNALIGNED_ACCESS
+	
+	if (mono_gc_is_moving ()) {
+		g_assert (size == mono_class_value_size (klass, NULL));
+		MONO_ENTER_NO_SAFEPOINTS;
+		gpointer data = mono_handle_get_data_unsafe (res_handle);
+		mono_gc_wbarrier_value_copy_internal (data, value, 1, klass);
+		MONO_EXIT_NO_SAFEPOINTS;
+	} else {
+		MONO_ENTER_NO_SAFEPOINTS;
+		gpointer data = mono_handle_get_data_unsafe (res_handle);
+#if NO_UNALIGNED_ACCESS
+		mono_gc_memmove_atomic (data, value, size);
+#else
+		switch (size) {
+		case 1:
+			*(guint8*)data = *(guint8 *) value;
+			break;
+		case 2:
+			*(guint16 *)(data) = *(guint16 *) value;
+			break;
+		case 4:
+			*(guint32 *)(data) = *(guint32 *) value;
+			break;
+		case 8:
+			*(guint64 *)(data) = *(guint64 *) value;
+			break;
+		default:
 			mono_gc_memmove_atomic (data, value, size);
-	#else
-			switch (size) {
-			case 1:
-				*(guint8*)data = *(guint8 *) value;
-				break;
-			case 2:
-				*(guint16 *)(data) = *(guint16 *) value;
-				break;
-			case 4:
-				*(guint32 *)(data) = *(guint32 *) value;
-				break;
-			case 8:
-				*(guint64 *)(data) = *(guint64 *) value;
-				break;
-			default:
-				mono_gc_memmove_atomic (data, value, size);
-			}
-	#endif
-			MONO_EXIT_NO_SAFEPOINTS;
 		}
+#endif
+		MONO_EXIT_NO_SAFEPOINTS;
 	}
+
 	if (m_class_has_finalize (klass))
 		mono_object_register_finalizer_handle (res_handle);
 

--- a/mono/metadata/object.c
+++ b/mono/metadata/object.c
@@ -6912,37 +6912,38 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	return_val_if_nok (error, NULL_HANDLE);
 
 	size -= MONO_ABI_SIZEOF (MonoObject);
-
-	if (mono_gc_is_moving ()) {
-		g_assert (size == mono_class_value_size (klass, NULL));
-		MONO_ENTER_NO_SAFEPOINTS;
-		gpointer data = mono_handle_get_data_unsafe (res_handle);
-		mono_gc_wbarrier_value_copy_internal (data, value, 1, klass);
-		MONO_EXIT_NO_SAFEPOINTS;
-	} else {
-		MONO_ENTER_NO_SAFEPOINTS;
-		gpointer data = mono_handle_get_data_unsafe (res_handle);
-#if NO_UNALIGNED_ACCESS
-		mono_gc_memmove_atomic (data, value, size);
-#else
-		switch (size) {
-		case 1:
-			*(guint8*)data = *(guint8 *) value;
-			break;
-		case 2:
-			*(guint16 *)(data) = *(guint16 *) value;
-			break;
-		case 4:
-			*(guint32 *)(data) = *(guint32 *) value;
-			break;
-		case 8:
-			*(guint64 *)(data) = *(guint64 *) value;
-			break;
-		default:
+	if (value != NULL) { 
+		if (mono_gc_is_moving ()) {
+			g_assert (size == mono_class_value_size (klass, NULL));
+			MONO_ENTER_NO_SAFEPOINTS;
+			gpointer data = mono_handle_get_data_unsafe (res_handle);
+			mono_gc_wbarrier_value_copy_internal (data, value, 1, klass);
+			MONO_EXIT_NO_SAFEPOINTS;
+		} else {
+			MONO_ENTER_NO_SAFEPOINTS;
+			gpointer data = mono_handle_get_data_unsafe (res_handle);
+	#if NO_UNALIGNED_ACCESS
 			mono_gc_memmove_atomic (data, value, size);
+	#else
+			switch (size) {
+			case 1:
+				*(guint8*)data = *(guint8 *) value;
+				break;
+			case 2:
+				*(guint16 *)(data) = *(guint16 *) value;
+				break;
+			case 4:
+				*(guint32 *)(data) = *(guint32 *) value;
+				break;
+			case 8:
+				*(guint64 *)(data) = *(guint64 *) value;
+				break;
+			default:
+				mono_gc_memmove_atomic (data, value, size);
+			}
+	#endif
+			MONO_EXIT_NO_SAFEPOINTS;
 		}
-#endif
-		MONO_EXIT_NO_SAFEPOINTS;
 	}
 	if (m_class_has_finalize (klass))
 		mono_object_register_finalizer_handle (res_handle);

--- a/mono/metadata/object.c
+++ b/mono/metadata/object.c
@@ -6907,7 +6907,6 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	if (mono_class_is_nullable (klass))
 		return mono_nullable_box_handle (value, klass, error);
 
-	
 	vtable = mono_class_vtable_checked (domain, klass, error);
 	return_val_if_nok (error, NULL_HANDLE);
 
@@ -6917,7 +6916,6 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 	return_val_if_nok (error, NULL_HANDLE);
 
 	size -= MONO_ABI_SIZEOF (MonoObject);
-	
 	if (mono_gc_is_moving ()) {
 		g_assert (size == mono_class_value_size (klass, NULL));
 		MONO_ENTER_NO_SAFEPOINTS;
@@ -6949,7 +6947,6 @@ mono_value_box_handle (MonoDomain *domain, MonoClass *klass, gpointer value, Mon
 #endif
 		MONO_EXIT_NO_SAFEPOINTS;
 	}
-
 	if (m_class_has_finalize (klass))
 		mono_object_register_finalizer_handle (res_handle);
 

